src/freertos_drivers/esp32/Esp32BootloaderHal.hxx

```diff
@@ -0,0 +1,562 @@
+
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(4,3,0)
+#error ESP32 Bootloader is only supported with ESP-IDF v4.3+
+#endif // IDF v4.3+
+
+#ifndef BOOTLOADER_LOG_LEVEL
+#define BOOTLOADER_LOG_LEVEL VERBOSE
+#endif // BOOTLOADER_LOG_LEVEL
+
+#ifndef BOOTLOADER_TWAI_LOG_LEVEL
+#define BOOTLOADER_TWAI_LOG_LEVEL VERBOSE
+#endif // BOOTLOADER_TWAI_LOG_LEVEL
+
+// Enable streaming support for the bootloader
+#define BOOTLOADER_STREAM
+// Set the buffer size to half the sector size to minimize the flash writes.
+#define WRITE_BUFFER_SIZE (CONFIG_WL_SECTOR_SIZE / 2)
+
+#include <driver/twai.h>
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4,4,0)
+#include <esp_chip_info.h>
+#endif // IDF v4.4+
+
+#include <esp_ota_ops.h>
+#if defined(CONFIG_IDF_TARGET_ESP32)
+#include <esp32/rom/rtc.h>
+#elif defined(CONFIG_IDF_TARGET_ESP32S2)
+#include <esp32s2/rom/rtc.h>
+#elif defined(CONFIG_IDF_TARGET_ESP32S3)
+#include <esp32s3/rom/rtc.h>
+#elif defined(CONFIG_IDF_TARGET_ESP32C3)
+#include <esp32c3/rom/rtc.h>
+#else
+#error Unknown/Unsupported ESP32 variant.
+#endif
+#include "openlcb/Bootloader.hxx"
+#include "utils/constants.hxx"
+#include "utils/Hub.hxx"
+
+/// ESP32 Bootloader internal data.
+struct Esp32BootloaderState
+{
+    /// Chip identifier for the currently running firmware.
+    esp_chip_id_t chip_id;
+
+    /// Currently running application header information.
+    ///
+    /// NOTE: Only the size is populated in this structure, the checksum values are
+    /// initialized to zero as they are not used.
+    struct app_header app_header;
+
+    /// Node ID to use for the bootloader.
+    uint64_t node_id;
+
+    /// Partition where the firmware is currently running from.
+    esp_partition_t *current;
+
+    /// Partition where the new firmware should be written to.
+    esp_partition_t *target;
+
+    /// OTA handle used to track the firmware update progress.
+    esp_ota_handle_t ota_handle;
+
+    /// Internal flag to indicate that we have initialized the TWAI peripheral and
+    /// should deinit it before exit.
+    bool twai_initialized;
+
+    /// GPIO pin connected to the CAN transceiver TX pin.
+    gpio_num_t tx_pin;
+
+    /// GPIO pin connected to the CAN transceiver RX pin.
+    gpio_num_t rx_pin;
+};
+
+/// Bootloader configuration data.
+static Esp32BootloaderState esp_bl_state;
+
+/// Maximum time to wait for a TWAI frame to be received or transmitted before
+/// giving up.
+static constexpr BaseType_t MAX_TWAI_WAIT = pdMS_TO_TICKS(250);
+
+/// Flag used to indicate that we have been requested to enter the bootloader
+/// instead of normal node operations. Note that this value will not be
+/// initialized by the system and a check for power on reset will need to be
+/// made to initialize it on first boot.
+static uint32_t RTC_NOINIT_ATTR bootloader_request;
+
+/// Value to be assigned to @ref bootloader_request when the bootloader should
+/// run instead of normal node operations.
+static constexpr uint32_t RTC_BOOL_TRUE = 1;
```

Comment:
Sure, it can be anything. I'd also probably suggest a rename on both `RTC_BOOL` constants to `ESP_BL_RTC_BOOL` to avoid potential global namespace clashes or can you think of another option?

Reply:
https://github.com/bakerstu/openmrn/pull/719